jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ConnectionSettingsMenus.kt

```diff
@@ -0,0 +1,200 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.core.credentials
+
+import com.intellij.icons.AllIcons
+import com.intellij.ide.DataManager
+import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.actionSystem.DefaultActionGroup
+import com.intellij.openapi.actionSystem.Presentation
+import com.intellij.openapi.actionSystem.ex.ComboBoxAction
+import com.intellij.openapi.project.DumbAware
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.ui.popup.JBPopupFactory
+import com.intellij.ui.components.JBLabel
+import com.intellij.util.EventDispatcher
+import com.intellij.util.ui.components.BorderLayoutPanel
+import software.aws.toolkits.core.credentials.CredentialIdentifier
+import software.aws.toolkits.core.region.AwsRegion
+import software.aws.toolkits.jetbrains.core.credentials.ChangeSettingsMode.BOTH
+import software.aws.toolkits.jetbrains.core.credentials.ChangeSettingsMode.CREDENTIALS
+import software.aws.toolkits.jetbrains.core.credentials.ChangeSettingsMode.REGIONS
+import software.aws.toolkits.jetbrains.core.credentials.ConnectionSettingsMenuBuilder.Companion.connectionSettingsMenuBuilder
+import software.aws.toolkits.resources.message
+import java.awt.BorderLayout
+import java.awt.event.MouseAdapter
+import java.awt.event.MouseEvent
+import javax.swing.JComponent
+import javax.swing.event.ChangeEvent
+import javax.swing.event.ChangeListener
+
+/**
+ * Determine what settings the settings selector is capable of changing
+ */
+enum class ChangeSettingsMode(val showRegions: Boolean, val showCredentials: Boolean) {
+    CREDENTIALS(false, true),
+    REGIONS(true, false),
+    BOTH(true, true)
+}
+
+/**
+ * Base logic for different ways to present connection settings in a consistent manner across the IDE
+ *
+ * @see ProjectLevelSettingSelector
+ * @see SettingsSelectorComboBoxAction
+ */
+
+// TODO: Write tests for these!
+abstract class SettingsSelectorLogicBase(private val menuMode: ChangeSettingsMode) {
+    private val listeners by lazy {
+        EventDispatcher.create(ChangeListener::class.java)
+    }
+
+    fun displayValue(): String = when (menuMode) {
+        CREDENTIALS -> credentialsDisplay()
+        REGIONS -> regionDisplay()
+        BOTH -> "${credentialsDisplay()}@${regionDisplay()}"
+    }
+
+    fun tooltip(): String? = when (menuMode) {
+        CREDENTIALS -> credentialsTooltip()
+        REGIONS -> regionTooltip()
+        BOTH -> null
```

Comment:
am surprised this lacks a tooltip

Reply:
Ya, current one does so it was kept that way. Not sure how I feel about it though...